internal/webhook/template_webhook.go

```diff
@@ -185,11 +190,134 @@ func (*ProviderTemplateValidator) ValidateUpdate(_ context.Context, _, _ runtime
+	}
 	return nil, nil
 }
 
 // Default implements webhook.Defaulter so a webhook will be registered for the type.
 func (*ProviderTemplateValidator) Default(_ context.Context, _ runtime.Object) error {
 	return nil
 }
+
+func (v TemplateValidator) templateIsInUse(ctx context.Context, template client.Object) (bool, admission.Warnings, error) {
+	inUseByCluster, err := v.templateIsInUseByCluster(ctx, template)
+	if err != nil {
+		return false, nil, err
+	}
+	if inUseByCluster {
+		return true, admission.Warnings{fmt.Sprintf("The %s object can't be removed if ManagedCluster objects referencing it still exist", v.templateKind)}, nil
+	}
+	inUseByChain, err := v.templateIsInUseByTemplateChain(ctx, template)
+	if err != nil {
+		return false, nil, err
+	}
+	if inUseByChain {
+		return true, admission.Warnings{fmt.Sprintf("The %s object can't be removed if %s object referencing it exists", v.templateKind, v.templateChainKind)}, nil
+	}
+	return false, nil, nil
+}
+
+func (v TemplateValidator) templateIsInUseByCluster(ctx context.Context, template client.Object) (bool, error) {
+	var key string
+
+	switch v.templateKind {
+	case v1alpha1.ClusterTemplateKind:
+		key = v1alpha1.ManagedClusterTemplateIndexKey
+	case v1alpha1.ServiceTemplateKind:
+		key = v1alpha1.ManagedClusterServiceTemplatesIndexKey
+	default:
+		return false, fmt.Errorf("invalid Template kind %s. Supported values are: %s and %s", v.templateKind, v1alpha1.ClusterTemplateKind, v1alpha1.ServiceTemplateKind)
+	}
+
+	managedClusters := &v1alpha1.ManagedClusterList{}
+	if err := v.Client.List(ctx, managedClusters,
+		client.InNamespace(template.GetNamespace()),
+		client.MatchingFields{key: template.GetName()},
+		client.Limit(1)); err != nil {
+		return false, err
+	}
+	if len(managedClusters.Items) > 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (v TemplateValidator) templateIsInUseByTemplateChain(ctx context.Context, template client.Object) (bool, error) {
+	listOpts := []client.ListOption{
+		client.InNamespace(template.GetNamespace()),
+		client.MatchingFields{v1alpha1.TemplateChainSupportedTemplatesIndexKey: template.GetName()},
+		client.Limit(1),
+	}
+	if v.templateChainKind == v1alpha1.ClusterTemplateChainKind {
+		chainList := &v1alpha1.ClusterTemplateChainList{}
+		if err := v.Client.List(ctx, chainList, listOpts...); err != nil {
+			return false, err
+		}
+		if len(chainList.Items) > 0 {
+			return true, nil
+		}
+	}
+	if v.templateChainKind == v1alpha1.ServiceTemplateChainKind {
+		chainList := &v1alpha1.ServiceTemplateChainList{}
+		if err := v.Client.List(ctx, chainList, listOpts...); err != nil {
+			return false, err
+		}
+		if len(chainList.Items) > 0 {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (v TemplateValidator) providerTemplateIsInUse(ctx context.Context, templateName string) (bool, error) {
```

Comment:
I think in our terms "provider template is in use" == it's part of some `Release` or it's explicitly set in the `Management` object